sklearn/cluster/_k_means_elkan.pyx

```diff
@@ -0,0 +1,140 @@
+# cython: cdivision=True
+# cython: boundscheck=False
+# cython: wraparound=False
+#
+# Author: Andreas Mueller
+#
+# Licence: BSD 3 clause
+
+import numpy as np
+cimport numpy as np
+cimport cython
+
+from libc.math cimport sqrt
+
+from ..metrics import euclidean_distances
+from ._k_means import _centers_dense
+
+
+cdef double *getfloatpointer(np.ndarray[np.float64_t, ndim=2, mode='c'] data):
+    return <double *>(data.data) 
+
+
+cdef double d(double* a, double* b, int n_features) nogil:
+    cdef double result, tmp
+    result = 0
+    cdef int i
+    for i in range(n_features):
+        tmp = (a[0] - b[0])
+        a += 1
+        b += 1
+        result += tmp * tmp
+    return sqrt(result)
+
+cdef assign_labels(double* X, double* centers, double[:, :]
+                   center_distances, int[:] labels, double[:, :] lower_bounds, double[:] distances, int n_samples, int n_features, int n_clusters):
+    # assigns closest center to X
+    # uses triangle inequality
+    cdef double* x
+    cdef double* c
+    cdef double d_c, dist
+    cdef int c_x, j, sample
+    for sample in range(n_samples):
+        # assign first cluster center
+        c_x = 0
+        x = X + sample * n_features
+        d_c = d(x, centers, n_features)
+        lower_bounds[sample, 0] = d_c
+        for j in range(n_clusters):
+            if d_c > center_distances[c_x, j]:
+                c = centers + j * n_features
+                dist = d(x, c, n_features)
+                lower_bounds[sample, j] = dist
+                if dist < d_c:
+                    d_c = dist
+                    c_x = j
+        labels[sample] = c_x
+        distances[sample] = d_c
+
+
+def k_means_elkan(X_, int n_clusters, init, float tol=1e-4, int max_iter=30, verbose=False):
```

Comment:
It seems that this function is meant to be used from outside. Thus I think that it should have a docstring.